### PR TITLE
Enable Achievements for Cata

### PR DESCRIPTION
**Pull request by aaron02**
**Description**
Fixed Achievements for cataclysm

**Todo / Checklist**
- [x] Target is branch develop.
- [x] Detailed description about this pull request.
- [x] Checked AE-Coding standards.

**Tests Performed:** 
- [x] Build AE.
- [x] Server startup.
- [x] Log into world.

### CODE REVIEW

File: src/world/GameCata/Storage/DBCStructures.h [hunk 1]
```diff
@@ -384,11 +384,18 @@ namespace DBC::Structures
...
-                uint32_t unused;                                    // 3
-                uint32_t questCount;                                // 4
+                uint32 currency;
+                uint32 count;
+            } currencyGain;
+
+            // ACHIEVEMENT_CRITERIA_TYPE_COMPLETE_DAILY_QUEST   = 14
+            struct
+            {
+                uint32  unused;                                 // 3
```
Comment (schnek) [inline src/world/GameCata/Storage/DBCStructures.h:397]: uint32_t

File: src/world/GameCata/Storage/DBCStructures.h [hunk 2]
```diff
@@ -403,6 +410,12 @@ namespace DBC::Structures
                 uint32_t mapID;                                     // 3
             } death_at_map;
 
+            // ACHIEVEMENT_CRITERIA_TYPE_DEATH_IN_DUNGEON       = 18
+            struct
+            {
+                uint32  manLimit;                               // 3
```
Comment (schnek) [inline src/world/GameCata/Storage/DBCStructures.h:416]: uint32_t

File: src/world/GameCata/Storage/DBCStructures.h [hunk 3]
```diff
@@ -422,6 +435,12 @@ namespace DBC::Structures
                 uint32_t fallHeight;                                // 4
             } fall_without_dying;
 
+            // ACHIEVEMENT_CRITERIA_TYPE_DEATHS_FROM            = 26
+            struct
+            {
+                uint32 type;                                    // 3, see enum EnviromentalDamage
```
Comment (schnek) [inline src/world/GameCata/Storage/DBCStructures.h:441]: uint32_t

File: src/world/GameCata/Storage/DBCStructures.h [hunk 4]
```diff
@@ -438,12 +457,20 @@ namespace DBC::Structures
...
             struct
             {
                 uint32_t spellID;                                   // 3
                 uint32_t castCount;                                 // 4
             } cast_spell;
 
+            // ACHIEVEMENT_CRITERIA_TYPE_BG_OBJECTIVE_CAPTURE
+            struct
+            {
+                uint32 objectiveId;                             // 3
```
Comment (schnek) [inline src/world/GameCata/Storage/DBCStructures.h:470]: uint32_t

File: src/world/GameCata/Storage/DBCStructures.h [hunk 5]
```diff
@@ -497,6 +524,13 @@ namespace DBC::Structures
                 uint32_t teamrating;                                // 4
             } reach_team_rating;
 
+            // ACHIEVEMENT_CRITERIA_TYPE_HIGHEST_PERSONAL_RATING = 39
+            struct
+            {
+                uint32 teamtype; // 3 {2, 3, 5}
```
Comment (schnek) [inline src/world/GameCata/Storage/DBCStructures.h:530]: uint32_t

File: src/world/Management/AchievementMgr.h [hunk 6]
```diff
@@ -197,9 +208,34 @@ enum AchievementCriteriaTypes
...
+    ACHIEVEMENT_CRITERIA_TYPE_BUY_GUILD_TABARD = 133,
+    ACHIEVEMENT_CRITERIA_TYPE_COMPLETE_QUESTS_GUILD = 134,
+    ACHIEVEMENT_CRITERIA_TYPE_HONORABLE_KILLS_GUILD = 135,
+    ACHIEVEMENT_CRITERIA_TYPE_KILL_CREATURE_TYPE_GUILD = 136,
+    ACHIEVEMENT_CRITERIA_TYPE_COMPLETE_GUILD_CHALLENGE_TYPE = 138, //struct { Flag flag; uint32 count; } 1: Guild Dungeon, 2:Guild Challenge, 3:Guild battlefield
+    ACHIEVEMENT_CRITERIA_TYPE_COMPLETE_GUILD_CHALLENGE = 139,  //struct { uint32 count; } Guild Challenge
+#endif
 };
 
+#if VERSION_STRING == WotLK
```
Comment (schnek) [inline src/world/Management/AchievementMgr.h:231]: why not create a separate file with macros
\world\Macros\AchievementMacros.hpp

File: src/world/Management/AchievementMgr.h [hunk 7]
```diff
@@ -197,9 +208,34 @@ enum AchievementCriteriaTypes
...
+    ACHIEVEMENT_CRITERIA_TYPE_CATCH_FROM_POOL = 127,
+    ACHIEVEMENT_CRITERIA_TYPE_BUY_GUILD_BANK_SLOTS = 128,
+    ACHIEVEMENT_CRITERIA_TYPE_EARN_GUILD_ACHIEVEMENT_POINTS = 129,
+    ACHIEVEMENT_CRITERIA_TYPE_WIN_RATED_BATTLEGROUND = 130,
+    ACHIEVEMENT_CRITERIA_TYPE_REACH_BG_RATING = 132,
+    ACHIEVEMENT_CRITERIA_TYPE_BUY_GUILD_TABARD = 133,
+    ACHIEVEMENT_CRITERIA_TYPE_COMPLETE_QUESTS_GUILD = 134,
+    ACHIEVEMENT_CRITERIA_TYPE_HONORABLE_KILLS_GUILD = 135,
+    ACHIEVEMENT_CRITERIA_TYPE_KILL_CREATURE_TYPE_GUILD = 136,
+    ACHIEVEMENT_CRITERIA_TYPE_COMPLETE_GUILD_CHALLENGE_TYPE = 138, //struct { Flag flag; uint32 count; } 1: Guild Dungeon, 2:Guild Challenge, 3:Guild battlefield
```
Comment (schnek) [inline src/world/Management/AchievementMgr.h:226]: code // struct - one tab at a time well, it just looks better ✨

File: src/world/Management/AchievementMgr.cpp [hunk 8]
```diff
@@ -1590,9 +1590,215 @@ void AchievementMgr::CompletedAchievement(DBC::Structures::AchievementEntry cons
...
+        data.writeBit(guid[4]);
+        data.writeBit(guid[2]);
+        data.writeBit(counter[6]);
+        data.writeBit(guid[0]);
+        data.writeBit(counter[4]);
+        data.writeBit(counter[1]);
+        data.writeBit(counter[2]);
+        data.writeBit(guid[3]);
+        data.writeBit(guid[7]);
+        data.writeBits(0, 2);           // criteria progress flags
```
Comment (schnek) [inline src/world/Management/AchievementMgr.cpp:1742]: code // struct - one tab at a time well, it just looks better ✨

File: src/world/Management/AchievementMgr.cpp [hunk 9]
```diff
@@ -1590,9 +1590,215 @@ void AchievementMgr::CompletedAchievement(DBC::Structures::AchievementEntry cons
...
+        data.writeBit(counter[0]);
+        data.writeBit(guid[5]);
+        data.writeBit(guid[6]);
+        data.writeBit(counter[7]);
+
+        criteriaData.WriteByteSeq(guid[3]);
+        criteriaData.WriteByteSeq(counter[4]);
+        criteriaData << uint32_t(0);      // timer 1
```
Comment (schnek) [inline src/world/Management/AchievementMgr.cpp:1750]: code // struct - one tab at a time well, it just looks better ✨

File: src/world/Management/AchievementMgr.cpp [hunk 10]
```diff
@@ -1590,9 +1590,215 @@ void AchievementMgr::CompletedAchievement(DBC::Structures::AchievementEntry cons
...
+        criteriaData.WriteByteSeq(counter[7]);
+        criteriaData.WriteByteSeq(guid[5]);
+        criteriaData.WriteByteSeq(counter[0]);
+        criteriaData.WriteByteSeq(guid[4]);
+        criteriaData.WriteByteSeq(guid[2]);
+        criteriaData.WriteByteSeq(guid[6]);
+        criteriaData.WriteByteSeq(guid[7]);
+        criteriaData.WriteByteSeq(counter[6]);
+        criteriaData << uint32_t(progressIter.first);
+        criteriaData << uint32_t(0);      // timer 2
```
Comment (schnek) [inline src/world/Management/AchievementMgr.cpp:1763]: code // struct - one tab at a time well, it just looks better ✨

**Review by schnek — CHANGES_REQUESTED** [review]

File: src/world/Management/AchievementMgr.cpp [hunk 11]
```diff
@@ -1590,9 +1590,213 @@ void AchievementMgr::CompletedAchievement(DBC::Structures::AchievementEntry cons
 
 //////////////////////////////////////////////////////////////////////////////////////////
 /// Sends all achievement data to the player. Also used for achievement inspection.
+#if VERSION_STRING == Cata
```
Comment (schnek) [inline src/world/Management/AchievementMgr.cpp:1593]: #if VERSION_STRING >= Cata

File: src/world/Management/AchievementMgr.cpp [hunk 12]
```diff
@@ -1590,9 +1590,213 @@ void AchievementMgr::CompletedAchievement(DBC::Structures::AchievementEntry cons
...
+    data.writeBits(m_completedAchievements.size(), 23);
+    data.flushBits();
+    data.append(criteriaData);
+
+    for (auto completeIter : m_completedAchievements)
+    {
+        if (!isVisible(completeIter))
+            continue;
+
+        data << uint32(completeIter.first);
```
Comment (Appled) [inline src/world/Management/AchievementMgr.cpp:1682]: uint32_t

File: src/world/Management/AchievementMgr.h [hunk 13]
```diff
@@ -197,9 +208,34 @@ enum AchievementCriteriaTypes
     ACHIEVEMENT_CRITERIA_TYPE_EARN_HONORABLE_KILL = 113,
     ACHIEVEMENT_CRITERIA_TYPE_ACCEPTED_SUMMONINGS = 114,
     ///\todo 0..114 => 124 criteria types total : Alleycat - We need to get the new Acheivements!
```
Comment (Appled) [inline src/world/Management/AchievementMgr.h:210]: I think you can remove this old comment as the types are there now

File: src/world/Management/AchievementMgr.h [hunk 14]
```diff
@@ -197,9 +208,34 @@ enum AchievementCriteriaTypes
...
+    ACHIEVEMENT_CRITERIA_TYPE_COMPLETE_GUILD_CHALLENGE_TYPE = 138, //struct { Flag flag; uint32 count; } 1: Guild Dungeon, 2:Guild Challenge, 3:Guild battlefield
+    ACHIEVEMENT_CRITERIA_TYPE_COMPLETE_GUILD_CHALLENGE = 139,  //struct { uint32 count; } Guild Challenge
+#endif
 };
 
+#if VERSION_STRING == WotLK
+    #define ACHIEVEMENT_CRITERIA_TYPE_TOTAL 124
+#endif
+#if VERSION_STRING > WotLK
+    #define ACHIEVEMENT_CRITERIA_TYPE_TOTAL 140
```
Comment (Appled) [inline src/world/Management/AchievementMgr.h:235]: Remove both of these macros and declare them inside the enum instead

File: src/world/Management/AchievementMgr.h [hunk 15]
```diff
@@ -261,6 +297,9 @@ class SERVER_DECL AchievementMgr
     void SaveToDB(QueryBuffer* buf);
     void CheckAllAchievementCriteria();
     void SendAllAchievementData(Player* player);
+#if VERSION_STRING >= Cata
+    void SendRespondInspectAchievements(Player* player); // Used when Inspecting a Player
```
Comment (Appled) [inline src/world/Management/AchievementMgr.h:301]: Methods should start with lowercase character

File: src/world/Objects/ObjectMgr.cpp [hunk 16]
```diff
@@ -1001,16 +1001,20 @@ AchievementCriteriaEntryList const & ObjectMgr::GetAchievementCriteriaByType(Ach
...
-#if VERSION_STRING < Cata
     for (uint32 rowId = 0; rowId < sAchievementCriteriaStore.GetNumRows(); ++rowId)
     {
         auto criteria = sAchievementCriteriaStore.LookupEntry(rowId);
         if (!criteria)
             continue;
 
-        m_AchievementCriteriasByType[criteria->requiredType].push_back(criteria);
-    }
+        auto achievement = sAchievementStore.LookupEntry(criteria->referredAchievement);
```
Comment (Appled) [inline src/world/Objects/ObjectMgr.cpp:1010]: Move this variable declaration inside the #if VERSION_STRING

**Review by Appled — CHANGES_REQUESTED** [review]

**Review by Appled — COMMENTED** [review]
You forgot one :P

**Review by Appled — APPROVED** [review]

**Pull request merged**